content/number-theory/ModLog.h

```diff
@@ -0,0 +1,26 @@
+
+/**
+ * Author: Andrei Navumenka, chilli
+ * Date: 2019-11-08
+ * License: Unlicense
+ * Source: https://github.com/indy256/codelibrary/blob/master/cpp/numbertheory/discrete_log.cpp
+ * Description: Calculates the discrete log. Returns an x s.t. a^x = b mod m. a and m must be coprime.
+ * Time: $O(n^{1/2})$
+ * Status: tested for all values under a 1000 for a,x, and m.
+ */
+#pragma once
+
+ll modLog(ll a, ll b, ll m) {
+	assert(__gcd(a, m) == 1);
+	ll n = (ll) sqrt(m) + 1, an = 1;
+	rep(i,0,n) an = an * a % m;
+	unordered_map<ll, ll> vals;
+	for (ll i = 1, x = an; i <= n; i++, x = x * an % m)
+		if (!vals.count(x)) vals[x] = i;
+	for (ll i = 0, x = b; i <= n; i++, x = x * a % m)
+		if (vals.count(x)) {
+			ll res = vals[x] * n - i;
+			if (res < m) return res;
```

Comment:
I don't see the point of this if statement; checking `< m` isn't enough to get a unique answer. Maybe we should take the minimum of all `res` instead, or add a comment in the description to that effect.

Reply:
I think it might be awkward if it returns a value more than `m`.

Reply:
But why? `m` is an unnatural thing to compare an exponent to, which lives mod phi(m). Taking the minimum seems like a good compromise to me -- it would be easy to explain and give a canonical output.

Reply:
Does the big step giant step algorithm necessarily enumerate through all possibilities for `x`?

Reply:
It should, and the `emplace` makes sure to only keep the first entry for each key, so we get the smallest value. Should clearly be tested though.

Reply:
I think I'd rather keep it as is - I don't think it's a big deal which value it returns.